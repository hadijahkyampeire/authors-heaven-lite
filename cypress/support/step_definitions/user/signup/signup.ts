/// <reference types="cypress" />
/// <reference types="../../../" />

import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import { currentUser } from '../../../utils';

Given("a user is on the landing page", () => {
  cy.visit("/");
});

When("a user clicks the register button", () => {
  cy.findByText('Register', { selector: "a" }).click();
});

Then('the user should go to the signup page', () => {
  cy.url().should('contain', '/register')
})

When("the user inputs information", () => {
  cy.get("#username").type('test1');
  cy.get("#email").type('test1@gmail.com');
  cy.get("#password").type('test123456');
});

Then("the user submits the form", () => {
  cy.get("form").submit();
});

Then("the user should go to login page", () => {
  // TODO this works on first run but I need to delete the user first
  // cy.url().should("contain", "/login");
});

