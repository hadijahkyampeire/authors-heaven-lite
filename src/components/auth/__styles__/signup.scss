$primary-green: #2ecc71;

.register-page {
  margin: auto;
  margin-top: 3rem;
  width: 40%;
  border: 1px dotted $primary-green;
  padding: 2rem;

  h2 {
    text-align: center;
    margin-bottom: 1rem;
  }
  form {
    margin-bottom: 1rem;
  }
  .action-buttons {
    display: flex;
    justify-content: space-between;

    .button {
      padding: 1rem 2.5rem;
      font-size: 1rem;
      outline: none;
    }
  }
  .cancel-login-btn {
    background-color: gray;
    color: #000;
  }
  .submit-login {
    background-color: $primary-green;
    padding: 1rem 2.5rem;
    font-size: 1rem;
  }

  .submit-login:focus {
    border-color: $primary-green !important;
    background-color: $primary-green;
    color: white;
    outline: 0 !important;
    box-shadow: 0 0 0 0 rgba(0, 0, 0, 0) !important;
  }

}
