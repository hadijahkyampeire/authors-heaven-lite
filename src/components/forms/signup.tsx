import React, { useState, ChangeEvent } from 'react';
import { Form, TextInput, Link } from 'carbon-components-react';
import { registerUser } from 'actions';
import { UserData } from 'types/user';

interface Props {
  registerUser: (
    ...args: Parameters<typeof registerUser>
  ) => void;
  user?: string;
};

export const SignUpForm: React.FC<Props> = ({ user, registerUser }) => {
  const initialValues: UserData = {
    email: "",
    username: "",
    password: ""
    
  };
  const [form, setForm] = useState<UserData>(initialValues);

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setForm({ ...form, [name]: value });
  };
  return (
    <Form onSubmit={(e) => { e.preventDefault(); registerUser(form); }} data-testid="signup-form">
    <div style={{marginBottom: '2rem'}}>
      <TextInput
        id="email"
        name="email"
        labelText="Email"
        placeholder="Enter Email"
        helperText="e.g example@gmail.com"
        value={form.email}
        onChange={handleInputChange}
      />
    </div>
    <div style={{marginBottom: '2rem'}}>
      <TextInput
        id="username"
        name="username"
        labelText="Username"
        placeholder="Enter Username"
        value={form.username}
        onChange={handleInputChange}
      />
    </div>
    <div style={{marginBottom: '2rem'}}>
      <TextInput
        helperText="Please choose a strong password with numbers, uppercase and special characters"
        id="password"
        name="password"
        invalidText="Invalid error message."
        labelText="Password"
        placeholder="Enter password"
        type="password"
        value={form.password}
        onChange={handleInputChange}
      />
  </div>
  <div className="action-buttons">
      <Link className="button cancel-login-btn">Cancel</Link>
      <button className="bx--btn bx--btn--primary submit-login" type="submit">Submit</button>
    </div>
</Form>
  );
};
