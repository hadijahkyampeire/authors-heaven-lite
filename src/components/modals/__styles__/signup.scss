$primary-green: #2ecc71;

.signup-modal {
  .bx--btn.bx--btn--primary {
    background-color: $primary-green;;
  }
}