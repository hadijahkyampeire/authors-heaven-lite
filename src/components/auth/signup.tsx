import React from 'react';
import { Redirect } from 'react-router';
import { SignUpForm } from 'components/forms';
import { Link } from 'carbon-components-react';
import { loginUser } from 'actions';
import './__styles__/signup.scss';

interface Props {
  registerUser: (
    ...args: Parameters<typeof loginUser>
  ) => void;
  username?: string;
};

export const SignUpPage: React.FC<Props> = ({ username, registerUser }) => {
  if (username) {
    <Redirect to='/login' />
  }

  return (
    <div className="register-page">
      <h2>REGISTER</h2>
      <SignUpForm user={username} registerUser={registerUser} />
      <span className="signup-message">Already have an account? <Link href="/login">Login</Link></span>
    </div>
  );
};
