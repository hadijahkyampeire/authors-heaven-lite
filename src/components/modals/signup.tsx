import React from 'react';
import { SignUpForm } from '../forms';
import { Modal } from "carbon-components-react";
import './__styles__/signup.scss';

interface Props {
  open: boolean,
  closeModal: () => void
}

const SignUpModal: React.FC<Props> = ({ open, closeModal }) => {
  return (
    <Modal
      className="signup-modal"
      modalHeading="SignUp"
      open={open}
      onRequestClose={closeModal}
      primaryButtonText="Register"
      secondaryButtonText="Cancel">
     <SignUpForm />
    </Modal>
  );
};

export { SignUpModal };