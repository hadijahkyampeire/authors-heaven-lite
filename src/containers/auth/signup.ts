import { connect } from 'react-redux';

import { registerUser } from 'actions';
import { SignUpPage} from 'components/auth';
import { AppState } from 'types';


const mapStateToProps = (state: AppState) => ({
  username: state.auth.user?.username,
});
const mapDispatchToProps = { registerUser };

export const Register = connect(
  mapStateToProps,
  mapDispatchToProps
)(SignUpPage);
