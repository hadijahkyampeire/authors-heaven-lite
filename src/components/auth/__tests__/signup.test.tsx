import React from "react";
import { BrowserRouter as Router } from "react-router-dom";

import { render } from "testUtils";

import  { SignUpPage } from "../signup";

type signUpPageProps = React.ComponentProps<typeof SignUpPage>;

const baseProps: signUpPageProps = {
  registerUser: jest.fn(),
  username: 'Haddy'
};

function renderUI(props: Partial<signUpPageProps> = {}) {
  return render(
    <Router>
      <SignUpPage {...baseProps} {...props} />
    </Router>
  );
};

describe("<SignUpPage />", () => {
  it("snapshot test", () => {
    const { container } = renderUI({});
    expect(container).toMatchSnapshot();
  });

  it("should display all elements of the page", async () => {
    const { getByText, findByTestId } = renderUI({});
    const signupForm = await findByTestId('signup-form');

    expect(getByText('REGISTER')).toBeInTheDocument();
    expect(signupForm).toBeInTheDocument();
    expect(getByText('Already have an account?')).toBeInTheDocument();
  });

  it("should have a form with input texts", async () => {
    const { findByTestId } = renderUI({});

    const signupForm = await findByTestId('signup-form');
    expect(signupForm).toHaveFormValues({
      email: "",
      username: "",
      password: ""
    });
  });
});
