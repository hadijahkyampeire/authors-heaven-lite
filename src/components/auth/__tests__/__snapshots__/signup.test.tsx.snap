// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUpPage /> snapshot test 1`] = `
<div>
  <div
    class="register-page"
  >
    <h2>
      REGISTER
    </h2>
    <form
      class="bx--form"
      data-testid="signup-form"
    >
       
      <div
        style="margin-bottom: 2rem;"
      >
        <div
          class="bx--form-item bx--text-input-wrapper"
        >
          <label
            class="bx--label"
            for="email"
          >
            Email
          </label>
          <div
            class="bx--text-input__field-outer-wrapper"
          >
            <div
              class="bx--text-input__field-wrapper"
            >
              <input
                class="bx--text-input bx--text__input"
                id="email"
                name="email"
                placeholder="Enter Email"
                title="Enter Email"
                type="text"
                value=""
              />
            </div>
            <div
              class="bx--form__helper-text"
            >
              e.g example@gmail.com
            </div>
          </div>
        </div>
      </div>
      <div
        style="margin-bottom: 2rem;"
      >
        <div
          class="bx--form-item bx--text-input-wrapper"
        >
          <label
            class="bx--label"
            for="username"
          >
            Username
          </label>
          <div
            class="bx--text-input__field-outer-wrapper"
          >
            <div
              class="bx--text-input__field-wrapper"
            >
              <input
                class="bx--text-input bx--text__input"
                id="username"
                name="username"
                placeholder="Enter Username"
                title="Enter Username"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        style="margin-bottom: 2rem;"
      >
        <div
          class="bx--form-item bx--text-input-wrapper"
        >
          <label
            class="bx--label"
            for="password"
          >
            Password
          </label>
          <div
            class="bx--text-input__field-outer-wrapper"
          >
            <div
              class="bx--text-input__field-wrapper"
            >
              <input
                class="bx--text-input bx--text__input"
                id="password"
                name="password"
                placeholder="Enter password"
                title="Enter password"
                type="password"
                value=""
              />
            </div>
            <div
              class="bx--form__helper-text"
            >
              Please choose a strong password with numbers, uppercase and special characters
            </div>
          </div>
        </div>
      </div>
      <div
        class="action-buttons"
      >
        <a
          class="bx--link button cancel-login-btn"
        >
          Cancel
        </a>
        <button
          class="bx--btn bx--btn--primary submit-login"
          type="submit"
        >
          Submit
        </button>
      </div>
       
    </form>
    <span
      class="signup-message"
    >
      Already have an account? 
      <a
        class="bx--link"
        href="/login"
      >
        Login
      </a>
    </span>
  </div>
</div>
`;
